File: haskell/glue/src/Glue/Lib/List/Last.hs
module Glue.Lib.List.Last where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

last :: IR Eval
last = NativeFunc lastImpl

lastImpl :: IR Eval -> Eval (IR Eval)
lastImpl arg = case arg of
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    List xs -> pure $ Prelude.last xs
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/last.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Last function - returns the last element of a list
/// Mirrors Haskell Glue.Lib.List.Last.last exactly
Ir last = IrNativeFunc(lastImpl);

/// Last function implementation
/// Mirrors Haskell Glue.Lib.List.Last.lastImpl exactly
Eval<Ir> lastImpl(Ir arg) {
  return switch (arg) {
    IrList(elements: final elements) =>
      elements.isNotEmpty
          ? Eval.pure(elements.last)
          : throwError(wrongArgumentType(['non-empty list'])),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
